### ISSUE TITLE
«internal Error» issue with 2-levels of nested expandable connectors vectors 

### ISSUE THREAD

**Issue by arouhan** [OpenModelica/OpenModelica#8247]
when using nested expandable connectors with 2 levels only, and the nested one is a vector of expandable connectors, I get an internal error.

I provide a case which highlight the issue. They are 2 models.
Both models have a top expandable connector which contains a sub-expandable connector. In the first model, the sub-expandable connector is an array of fixed dimension of an expandable connector.

Both models have various input and output to some variable in the top expandable connector and in the sub-expandable  connector. The first model writes to a variable to the array of expandable connector.

When I run the first model under OME v1.18, I got the following internal error:
` Error: Internal error NFConnection.split got unbalanced connection connect(ad1.hb, h) (lhs: 1, rhs: 2) `
I understand that according to @perost message in ticket #8094, having two levels of nested connectors should work with OME v1.18. Is this related to the #8094 issue which is solved as part of the next release ? or is this error unrelated to this ticket and is specific of the use of the array of expandable connectors?

The second model is the same as the first one with the sub-expandable connector a single one, ie, no vector. When I run the second model under OME v1.18, it checks and runs ok.

Both models checks ok and runs ok under Dymola 2022x.

[testNestedVectorExpandableConnectors.zip](https://github.com/OpenModelica/OpenModelica/files/7625571/testNestedVectorExpandableConnectors.zip)

**Comment by perost** [comment]
It's not fixed by #8094, though it improves the error message a bit:
```
Error: Internal error NFConnection.split got unbalanced connection connect(ad1.hb, h):
  lhs: {ad1.hb.position}
  rhs: {h.position, h.pb.speed}
```
Of course, this should never happen regardless. Both `ad1.hb` and `h` are expandable connectors, so by connecting them they should by definition contain the same elements.

**Comment by perost** [comment]
I had a closer look at this, and the issue is that we don't properly support arrays of expandable connectors. Having an array of expandable connectors means that the elements of the array might get different types depending on how they're connected, which we can't handle at the moment. We actually don't even support connecting only parts of regular arrays in expandable connectors yet, since it's not something that's really used normally.

Handling this properly would likely require rewriting a lot of the expandable connector handling, so I'm tentatively changing the milestone to 1.20.0 since it's not something that currently has very high priority and is unlikely to be fixed in 1.19.0.

**Comment by casella** [comment]
I agree. In fact, I do see the rationale for using nested expandable connectors, to manage complex systems avoiding the city dump effect on bus signal names. @arouhan, could you elaborate on the use case for nested arrays of expandable connectors?

**Comment by arouhan** [comment]
we are working with systems that have several sub-systems working in parallel. Typically, several engines and connected attachements (alternators, propellers, etc). This is typical in the industry to have several parallel trains. The number of these subsystems depends on design of the system. As we are building a library, we found more flexible to cope with an array of expandable connectors to collect and transmit signals. We could have build something like:
```
expandable connector EngineBus;

EngineBus engineBus1 ;
EngineBus engineBus2 ;
//etc.
```


that would mean we have a hard-coded predefined maximum number of subsystems, which can be limiting.
Intead, we prefer:
```
parameter Integer noEngines = 4 ;
EngineBus engineBus[noEngines]  ;
```
in our application, we use adaptors to connect the signal to the expandable connector, so that all possible signals that transit through the expandable connectors are known in advance without mistake possible in the naming. And consequently all the signals connected to a specifc expandable connector are identical to another instance of identical expendable connector type - or at least they belong to the same set of possible signals.


@perost can you elaborate on the following sentance, in particular, why the elements of the array would have different types, and what it does mean ?
> Having an array of expandable connectors means that the elements of the array might get different types


Just to share my understadning: while thinking using arrays of nested expandable connectors, I though:
```EngineBus engineBus[2];```
would be flattened as:
```
EngineBus engineBus[1];
EngineBus engineBus[2];
```

which is not so different from:
```
EngineBus engineBus1;
EngineBus engineBus2;
```
and apparently I'm wrong here.

**Comment by perost** [comment]
> @perost can you elaborate on the following sentance, in particular, why the elements of the array would have different types, and what it does mean ?
> 
> > Having an array of expandable connectors means that the elements of the array might get different types
> 
> Just to share my understadning: while thinking using arrays of nested expandable connectors, I though: `EngineBus engineBus[2];` would be flattened as:
> 
> ```
> EngineBus engineBus[1];
> EngineBus engineBus[2];
> ```
> 
> which is not so different from:
> 
> ```
> EngineBus engineBus1;
> EngineBus engineBus2;
> ```
> 
> and apparently I'm wrong here.

The issue is that we handle expandable connectors on an instance tree level, where `engineBus` is just one node with some dimensions. That works fine for scalars, but when you have arrays of expandable connectors it doesn't work since each element of the array will contain different components depending on how they're connected. Hence why we'll need to rewrite quite a lot of our expandable connection handling to be able to support this.

**Comment by casella** [comment]
@arouhan it seems that you are using OMC to develop complex industrial applications. If you need some specific features in the compiler with high priority, you may join the [Open Source Modelica Consortium](https://openmodelica.org/home/consortium) and we could possibly arrange a DFD contract to get that done within a given time frame. 

If you are interested, please contact me at my email francesco.casella@polimi.it
